pkg/crypto/certloader/client.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certloader
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+)
+
+// ClientConfig creates tls.Config to be used as TLS client.
+type ClientConfig interface {
+	IsMutualTLS() bool
+	ClientConfig(base *tls.Config) *tls.Config
+}
```

Comment:
An interface that is implemented by both a client configuration and server configuration is precisely what I wish to avoid.

At a high level, we're given either a client configuration or a server configuration (e.g. by cli flags). By constructing a `TLSConfigLoader` we're effectively loosing this information and open the door for misuse; a server configuration could be used as a client configuration, and conversely a client configuration could be used as a server configuration.

By having two incompatible interfaces we're preventing that type of bug at compile time by design. I see no gain in having them compatible.

Reply:
We discussed this offline and agreed on
```go
type ClientTLSConfig interface {
    IsMutualTLS() bool
    ClientTLSConfig(base *tls.Config) *tls.Config
}
```
And similarly for the server counterpart.